core/src/main/java/org/lflang/ast/MalleableString.java

```diff
@@ -348,40 +349,51 @@ private List<String> getLinesOfInterest(
     }
 
     @Override
-    public void findBestRepresentation(
+    public boolean findBestRepresentation(
         Supplier<RenderResult> providedRender,
         ToLongFunction<RenderResult> badness,
         int width,
         int indentation,
         String singleLineCommentPrefix) {
       this.width = width;
       keepCommentsOnSameLine = true;
-      optimizeChildren(providedRender, badness, width, indentation, singleLineCommentPrefix);
+      var everChanged = false;
+      var changed =
+          optimizeChildren(providedRender, badness, width, indentation, singleLineCommentPrefix);
+      everChanged = changed;
+      if (changed)
+        changed =
+            optimizeChildren(providedRender, badness, width, indentation, singleLineCommentPrefix);
       if (components.stream()
           .noneMatch(
               it ->
                   it.render(indentation, singleLineCommentPrefix, false, null)
                       .unplacedComments
                       .findAny()
-                      .isPresent())) return;
+                      .isPresent())) return changed;
       long badnessTrue = badness.applyAsLong(providedRender.get());
       keepCommentsOnSameLine = false;
-      optimizeChildren(providedRender, badness, width, indentation, singleLineCommentPrefix);
+      changed =
+          optimizeChildren(providedRender, badness, width, indentation, singleLineCommentPrefix);
+      everChanged |= changed;
       long badnessFalse = badness.applyAsLong(providedRender.get());
       keepCommentsOnSameLine = badnessTrue < badnessFalse;
-      optimizeChildren(providedRender, badness, width, indentation, singleLineCommentPrefix);
-      optimizeChildren(providedRender, badness, width, indentation, singleLineCommentPrefix);
+      if (changed)
+        changed =
+            optimizeChildren(providedRender, badness, width, indentation, singleLineCommentPrefix);
+      if (changed)
+        optimizeChildren(providedRender, badness, width, indentation, singleLineCommentPrefix);
```

Comment:
This call seems redundant, and you're not doing anything with the return value...

Reply:
These calls have side effects. Multiple calls to the same function with the same arguments may be required because as parts of the textual representation get updated, the optimal representation of other parts may change. in particular, if the text is wider than 100 characters, the line may only need to be broken in one place, but it will be broken in multiple places if this second optimization pass is not made.

This is a heuristic because in order to get a truly optimal textual representation it might be necessary to make combinatorially many calls with different combinations of textual representations.

Reply:
I see. There might not even be a fixed point? Perhaps a small comment with a hint would be useful because it does look a bit sus.

Reply:
Yes, no fixed point is guaranteed, but since a subsequent call to the formatter will get the same AST and therefore have the same starting point, the formatter as a whole should still be idempotent.